File: examples/hello-world/gabriel-greeting-ruby/internal/server_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../support"
require "grpc"
require "gen/holons/v1/manifest_pb"
require "gen/holons/v1/describe_pb"
require "gen/holons/v1/describe_services_pb"
require "v1/greeting_services_pb"
require_relative "server"

class GreetingServerTest < Minitest::Test
  def setup
    @server = GRPC::RpcServer.new
    GabrielGreetingRuby::Internal::Server.register_services(@server)
    @port = @server.add_http2_port("127.0.0.1:0", :this_port_is_insecure)
    @thread = Thread.new { @server.run_till_terminated }
    raise "gRPC server failed to start" unless @server.wait_till_running(5)

    target = "127.0.0.1:#{@port}"
    @greeting_stub = Greeting::V1::GreetingService::Stub.new(target, :this_channel_is_insecure, timeout: 5)
    @meta_stub = Holons::V1::HolonMeta::Stub.new(target, :this_channel_is_insecure, timeout: 5)
  end

  def teardown
    return if @server.nil?

    @server.stop if @server.respond_to?(:running?) && @server.running?
    @thread&.join(1)
  end

  def test_list_languages_returns_all_languages
    response = @greeting_stub.list_languages(Greeting::V1::ListLanguagesRequest.new)
    assert_equal 56, response.languages.length
  end

  def test_say_hello_uses_requested_language
    response = @greeting_stub.say_hello(
      Greeting::V1::SayHelloRequest.new(name: "Bob", lang_code: "fr")
    )

    assert_equal "Bonjour Bob", response.greeting
    assert_equal "French", response.language
    assert_equal "fr", response.lang_code
  end

  def test_say_hello_uses_localized_default_name
    response = @greeting_stub.say_hello(
      Greeting::V1::SayHelloRequest.new(lang_code: "fr")
    )

    assert_equal "Bonjour Marie", response.greeting
    assert_equal "fr", response.lang_code
  end

  def test_say_hello_emits_observability_signals
    old_op_obs = ENV["OP_OBS"]
    ENV["OP_OBS"] = "logs,metrics"
    Holons::Observability.reset
    obs = Holons::Observability.configure(
      Holons::Observability::Config.new(slug: "gabriel-greeting-ruby")
    )

    response = @greeting_stub.say_hello(
      Greeting::V1::SayHelloRequest.new(name: " Bob ", lang_code: "en")
    )

    assert_equal "Hello Bob", response.greeting
    counter = obs.registry.counters.find { |sample| sample.name == "greeting_emitted_total" }
    refute_nil counter
    assert_equal(
      {
        "lang_code" => "en",
        "language" => "English",
        "transport" => "unknown"
      },
      counter.labels
    )
    assert_equal 1, counter.value

    entry = obs.log_ring.drain.find { |log_entry| Holons::Observability.body_string(log_entry.record) == "Greeted Bob in English (en)" }
    refute_nil entry
    fields = Holons::Observability.attributes_hash(entry.record.attributes)
    assert_equal(
      %w[duration_ns greeting lang_code language name transport],
      fields.keys.sort
    )
    assert_equal "en", fields["lang_code"]
    assert_equal "English", fields["language"]
    assert_equal "Bob", fields["name"]
    assert_equal "Hello Bob", fields["greeting"]
    assert_equal "unknown", fields["transport"]
    assert_instance_of Integer, fields["duration_ns"]
    assert_operator fields["duration_ns"], :>=, 0
    assert_equal(
      "gabriel-greeting-ruby",
      Holons::Observability.attribute_string(entry.record.attributes, Holons::Observability::ATTR_HOLONS_SLUG)
    )
    assert_equal(
      "gabriel-greeting-ruby",
      Holons::Observability.attribute_string(entry.record.attributes, Holons::Observability::ATTR_SERVICE_NAME)
    )
    refute_empty Holons::Observability.attribute_string(entry.record.attributes, Holons::Observability::ATTR_HOLONS_SESSION_ID)
    duration_attr = Holons::Observability.attribute_value(entry.record.attributes, "duration_ns")
    assert_equal :int_value, duration_attr.value
  ensure
    Holons::Observability.reset
    if old_op_obs.nil?
      ENV.delete("OP_OBS")
    else
      ENV["OP_OBS"] = old_op_obs
    end
  end

  def test_say_hello_emits_stdio_transport_under_stdio_serve
    Holons::Observability.require_grpc_observability_support!
    env = ENV.to_h.merge(
      "OP_OBS" => "logs",
      "OP_INSTANCE_UID" => "ruby-stdio-fixture"
    )
    proxy = Holons::DiscoverySupport::StdioDialProxy.new(
      RbConfig.ruby,
      args: ["cmd/main.rb"],
      chdir: GabrielGreetingRuby::ROOT,
      env: env
    )
    proxy.start
    channel = Holons::DiscoverySupport.dial_ready(proxy.target, 5000)

    greeting_stub = Greeting::V1::GreetingService::Stub.new(
      "unused",
      :this_channel_is_insecure,
      channel_override: channel,
      timeout: 5
    )
    observability_stub = Holons::V1::HolonObservability::Stub.new(
      "unused",
      :this_channel_is_insecure,
      channel_override: channel,
      timeout: 5
    )

    response = greeting_stub.say_hello(
      Greeting::V1::SayHelloRequest.new(name: "Ana", lang_code: "es")
    )
    assert_equal "Hola Ana", response.greeting

    records = observability_stub.logs(Holons::V1::LogsRequest.new).to_a
    record = records.find { |candidate| Holons::Observability.body_string(candidate) == "Greeted Ana in Spanish (es)" }
    refute_nil record
    fields = Holons::Observability.attributes_hash(record.attributes)
    assert_equal "stdio", fields["transport"]
    assert_instance_of Integer, fields["duration_ns"]
    assert_equal :int_value, Holons::Observability.attribute_value(record.attributes, "duration_ns").value
  ensure
    Holons::DiscoverySupport.close_channel(channel) unless channel.nil?
    proxy&.close
  end

  def test_describe_uses_static_manifest
    response = @meta_stub.describe(Holons::V1::DescribeRequest.new)

    assert_equal "Gabriel", response.manifest.identity.given_name
    assert_equal "Greeting-Ruby", response.manifest.identity.family_name
    assert_equal "Greets users in 56 languages — a Ruby daemon example.", response.manifest.identity.motto
    assert_equal ["greeting.v1.GreetingService"], response.services.map(&:name)
  end

  def test_normalize_listen_uri_expands_empty_tcp_host
    assert_equal "tcp://0.0.0.0:9090", GabrielGreetingRuby::Internal::Server.normalize_listen_uri("tcp://:9090")
    assert_equal "tcp://127.0.0.1:9090", GabrielGreetingRuby::Internal::Server.normalize_listen_uri("tcp://127.0.0.1:9090")
  end
end

File: examples/hello-world/gabriel-greeting-python/_internal/server_test.py
from __future__ import annotations

import unittest
from concurrent import futures
from unittest.mock import patch

import grpc

from support import ensure_import_paths

ensure_import_paths()

from holons import describe, observability
from v1 import greeting_pb2, greeting_pb2_grpc

from _internal.server import GreetingService, normalize_listen_uri


class GreetingServerTest(unittest.TestCase):
    def setUp(self) -> None:
        self.server = grpc.server(futures.ThreadPoolExecutor(max_workers=2))
        greeting_pb2_grpc.add_GreetingServiceServicer_to_server(
            GreetingService(), self.server
        )
        port = self.server.add_insecure_port("127.0.0.1:0")
        self.server.start()
        self.channel = grpc.insecure_channel(f"127.0.0.1:{port}")
        grpc.channel_ready_future(self.channel).result(timeout=5)
        self.stub = greeting_pb2_grpc.GreetingServiceStub(self.channel)

    def tearDown(self) -> None:
        self.channel.close()
        self.server.stop(0).wait(timeout=5)

    def test_list_languages_returns_all_languages(self) -> None:
        response = self.stub.ListLanguages(greeting_pb2.ListLanguagesRequest(), timeout=5)
        self.assertEqual(len(response.languages), 56)

    def test_list_languages_populates_required_fields(self) -> None:
        response = self.stub.ListLanguages(greeting_pb2.ListLanguagesRequest(), timeout=5)
        for language in response.languages:
            self.assertTrue(language.code)
            self.assertTrue(language.name)
            self.assertTrue(language.native)

    def test_say_hello_uses_requested_language(self) -> None:
        response = self.stub.SayHello(
            greeting_pb2.SayHelloRequest(name="Bob", lang_code="fr"),
            timeout=5,
        )
        self.assertEqual(response.greeting, "Bonjour Bob")
        self.assertEqual(response.language, "French")
        self.assertEqual(response.lang_code, "fr")

    def test_say_hello_uses_localized_default_name(self) -> None:
        response = self.stub.SayHello(
            greeting_pb2.SayHelloRequest(lang_code="fr"),
            timeout=5,
        )
        self.assertEqual(response.greeting, "Bonjour Marie")
        self.assertEqual(response.lang_code, "fr")

    def test_say_hello_falls_back_to_english(self) -> None:
        response = self.stub.SayHello(
            greeting_pb2.SayHelloRequest(name="Bob", lang_code="xx"),
            timeout=5,
        )
        self.assertEqual(response.greeting, "Hello Bob")
        self.assertEqual(response.lang_code, "en")

    def test_say_hello_emits_observability_signals(self) -> None:
        self.addCleanup(observability.reset)
        observability.reset()
        with patch.dict("os.environ", {"OP_OBS": "logs,metrics"}):
            obs = observability.configure(
                observability.Config(slug="gabriel-greeting-python")
            )
            response = self.stub.SayHello(
                greeting_pb2.SayHelloRequest(name=" Bob ", lang_code="en"),
                timeout=5,
            )

        self.assertEqual(response.greeting, "Hello Bob")
        self.assertIsNotNone(obs.registry)
        snapshot = obs.registry.snapshot()
        counters = [
            counter
            for counter in snapshot["counters"]
            if counter[0] == "greeting_emitted_total"
        ]
        self.assertEqual(len(counters), 1)
        _name, _help, labels, value = counters[0]
        self.assertEqual(
            labels,
            {"lang_code": "en", "language": "English", "transport": "unknown"},
        )
        self.assertEqual(value, 1)

        self.assertIsNotNone(obs.log_ring)
        entries = [
            entry
            for entry in obs.log_ring.drain()
            if entry.message == "Greeted Bob in English (en)"
        ]
        self.assertEqual(len(entries), 1)
        fields = entries[0].fields
        self.assertEqual(
            set(fields),
            {"lang_code", "language", "name", "greeting", "transport", "duration_ns"},
        )
        self.assertEqual(fields["lang_code"], "en")
        self.assertEqual(fields["language"], "English")
        self.assertEqual(fields["name"], "Bob")
        self.assertEqual(fields["greeting"], "Hello Bob")
        self.assertEqual(fields["transport"], "unknown")
        self.assertGreaterEqual(int(fields["duration_ns"]), 0)

    def test_normalize_listen_uri_expands_empty_tcp_host(self) -> None:
        self.assertEqual(normalize_listen_uri("tcp://:9090"), "tcp://0.0.0.0:9090")
        self.assertEqual(normalize_listen_uri("tcp://127.0.0.1:9090"), "tcp://127.0.0.1:9090")

    def test_static_describe_response_is_registered(self) -> None:
        response = describe.static_response()
        self.assertEqual(response.manifest.identity.given_name, "Gabriel")
        self.assertEqual(response.manifest.lang, "python")
        self.assertEqual(response.manifest.build.runner, "python")


if __name__ == "__main__":
    unittest.main()
